compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -224,6 +224,157 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
+                    }
+                    _ => None,
+                };
+                chain_refutabilities.push(refutability);
+            };
+            if let hir::Expr {
+                kind: hir::ExprKind::Binary(Spanned { node: hir::BinOpKind::And, .. }, lhs, rhs),
+                ..
+            } = cur_expr
+            {
+                add(rhs);
+                cur_expr = lhs;
+            } else {
+                add(cur_expr);
+                break;
+            }
+        }
+        chain_refutabilities.reverse();
+
+        // Third, emit the actual warnings.
+
+        if chain_refutabilities.iter().all(|r| matches!(*r, Some((_, false)))) {
+            // The entire chain is made up of irrefutable `let` statements
+            let let_source = let_source_parent(self.tcx, top, None);
+            irrefutable_let_patterns(
+                cx.tcx,
+                top,
+                let_source,
+                chain_refutabilities.len(),
+                top_expr.span,
+            );
+            return true;
+        }
+        let lint_affix = |affix, kind, suggestion| {
+            let span_start = affix[0].unwrap().0;
+            let span_end = affix.last().unwrap().unwrap().0;
+            let span = span_start.to(*span_end);
+            let cnt = affix.len();
+            cx.tcx.struct_span_lint_hir(IRREFUTABLE_LET_PATTERNS, top, span, |lint| {
+                let mut diag = lint.build(&format!("{kind} of irrefutable patterns in let chain"));
+                diag.note(&format!(
+                    "{these} pattern{s} will always match",
+                    these = pluralize!("this", cnt),
+                    s = pluralize!(cnt)
+                ));
+                diag.help(&format!(
+                    "consider moving {} {suggestion}",
+                    if cnt > 1 { "them" } else { "it" }
+                ));
+                diag.emit()
+            });
+        };
+        if let Some(until) = chain_refutabilities.iter().position(|r| !matches!(*r, Some((_, false)))) && until > 0 {
+            // The chain has a non-zero prefix of irrefutable `let` statements.
+
+            // Check if the let source is while, for there is no alternative place to put a prefix,
+            // and we shouldn't lint.
+            let let_source = let_source_parent(self.tcx, top, None);
+            if !matches!(let_source, LetSource::WhileLet) {
+                // Emit the lint
+                let prefix = &chain_refutabilities[..until];
+                lint_affix(prefix, "prefix", "outside of the construct");
+            }
+        }
+        if let Some(from) = chain_refutabilities.iter().rposition(|r| !matches!(*r, Some((_, false)))) && from != (chain_refutabilities.len() - 1) {
+            // The chain has a non-empty suffix of irrefutable `let` statements
+            let suffix = &chain_refutabilities[from + 1..];
+            lint_affix(suffix, "suffix", "into the body");
```

Comment:
> Do we care enough to specify if there are a prefix or suffix?

There are different suggestions: either moving the patterns to the parent block, or moving them to the `if`'s body. Second, `leading` vs `trailing` (or prefix vs suffix) is easier to understand for readers than affix I think.

> What is the output in a case like `if let Some(x) = Some(Some(1)) && let y = x && let Some(_z) = y {}`?

Irrefutable patterns that are surrounded by refutable ones are allowed, because there is no low-cost resolution to the lint that you can always suggest. The straightforward resolution is to create nested let chains with the assignments in between, but this is not a low-cost resolution like it is to move the `let` into the body or surrounding body.

Yes, sometimes the inner irrefutable pattern might be something that can be reordered so that it is equivalent to have it in the body or surrounding body, but figuring out the dependencies is a hard problem, especially due to possible side effects in the computation.

Personally I found inner irrefutable patterns useful already despite my limited exposure to the `if_chain` macro.